File: tasks/01-run-length/python/rle.py
from itertools import groupby


def encode(s: str) -> str:
    """
    Run‑length encode the input string.

    >>> encode("AAB") -> "A2B1"
    """
    # TODO: implement
    # raise NotImplementedError("Implement me!")

    # raise NotImplementedError("Implement me!")
    # grouping the string like ("a", generator value) and then sum the gererator value
    return "".join(f"{char}{sum(1 for _ in grp)}" for char, grp in groupby(s))

# Solution : Using groupby to group the element with generator value and sum the value

File: tasks/01-run-length/go/rle.go
package rle

import (
	"strconv"
	"strings"
	"unicode/utf8"
)

// Encode returns the run‑length encoding of UTF‑8 string s.
//
// "AAB" → "A2B1"
func Encode(s string) string {
	// TODO: implement
	if len(s) == 0 {
		return ""
	}

	var result strings.Builder
	currentRune, _ := utf8.DecodeRuneInString(s)
	count := 1

	for _, r := range s[utf8.RuneLen(currentRune):] {
		if r == currentRune {
			count++
		} else {
			result.WriteRune(currentRune)
			result.WriteString(strconv.Itoa(count))
			currentRune = r
			count = 1
		}
	}
	result.WriteRune(currentRune)
	result.WriteString(strconv.Itoa(count))

	return result.String()
}
